core/src/main/java/com/linecorp/armeria/common/StringMultimap.java

```diff
@@ -0,0 +1,1098 @@
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterators;
+
+import io.netty.handler.codec.DateFormatter;
+import io.netty.util.AsciiString;
+
+/**
+ * The base container implementation of {@link HttpHeaders}, {@link HttpHeadersBuilder}, {@link QueryParams}
+ * and {@link QueryParamsBuilder}.
+ *
+ * @see HttpHeadersBase
+ * @see QueryParamsBase
+ */
+abstract class StringMultimap<IN_NAME extends CharSequence, NAME extends IN_NAME>
+        implements StringMultimapGetters<IN_NAME, NAME> {
+
+    static final int DEFAULT_SIZE_HINT = 16;
+
+    /**
+     * Constant used to seed the hash code generation. Could be anything but this was borrowed from murmur3.
+     */
+    static final int HASH_CODE_SEED = 0xc2b2ae35;
+
+    // XXX(anuraaga): It could be an interesting future optimization if we can use something similar
+    //                to an EnumSet when it's applicable, with just one each of commonly known header names.
+    //                It should be very common.
+    @VisibleForTesting
+    final Entry[] entries;
+    private final byte hashMask;
+
+    private final Entry firstGroupHead;
+    private Entry secondGroupHead;
+
+    int size;
+
+    StringMultimap(int sizeHint) {
+        // Enforce a bound of [2, 128] because hashMask is a byte. The max possible value of hashMask is
+        // one less than the length of this array, and we want the mask to be > 0.
+        @SuppressWarnings("unchecked")
+        final Entry[] newEntries = (Entry[]) Array.newInstance(
+                Entry.class, findNextPositivePowerOfTwo(max(2, min(sizeHint, 128))));
+        entries = newEntries;
+        hashMask = (byte) (entries.length - 1);
+        firstGroupHead = secondGroupHead = new Entry();
+    }
+
+    /**
+     * Creates a shallow or deep copy of the specified {@link StringMultimap}.
+     */
+    StringMultimap(StringMultimap<IN_NAME, NAME> parent, boolean shallowCopy) {
+        hashMask = parent.hashMask;
+
+        if (shallowCopy) {
+            entries = parent.entries;
+            firstGroupHead = parent.firstGroupHead;
+            secondGroupHead = parent.secondGroupHead;
+            size = parent.size;
+        } else {
+            @SuppressWarnings("unchecked")
+            final Entry[] newEntries = (Entry[]) Array.newInstance(Entry.class, parent.entries.length);
+            entries = newEntries;
+            firstGroupHead = secondGroupHead = new Entry();
+            final boolean succeeded = addFast(parent);
+            assert succeeded;
+        }
+    }
+
+    /**
+     * Creates a deep copy of the specified {@link StringMultimapGetters}.
+     */
+    StringMultimap(StringMultimapGetters<IN_NAME, NAME> parent) {
+        this(parent.size());
+        assert !(parent instanceof StringMultimap);
+        addSlow(parent);
+    }
+
+    // Extension points
+
+    abstract int hashName(IN_NAME name);
+
+    abstract boolean nameEquals(@Nullable NAME a, @Nullable IN_NAME b);
```

Comment:
If I'm not mistaken, I think it's possible to handle `null` in this class - then `nameEquals` doesn't need nullable parameters which is simpler for the subclasses.

By the way, I looked into this mainly since I noticed the subclasses handle two `null` arguments differently, headers returns `false` while query params returns `true` right now which was surprising.